In [12]:
import sys
sys.path.append('../')
import pandas as pd
import numpy as np
from xgboost import XGBRegressor
from datetime import timedelta
import pickle

from utils import utils_scenario as utils, data_preparation as prep, data_process as proc

In [13]:
#read in processed measurement data and forecast data
measurement = pd.read_csv('./data/processed/last_measurement.csv')
measurement['datetime']= pd.to_datetime(measurement['datetime'],format='%Y-%m-%d %H:%M:%S')
forecast = pd.read_csv('./data/processed/last_forecast.csv')
forecast['f_date']= pd.to_datetime(forecast['f_date'],format='%Y-%m-%d %H:%M:%S')
# forecast['p_date']= pd.to_datetime(forecast['p_date'],format='%Y-%m-%d %H:%M:%S')

In [14]:
#prepare processed data into same format as training df 
def prepare_x_test(measurement, forecast, past_n_steps, pred_period):
    
    #concat past n steps from measurement 
    df = measurement.set_index('datetime')
    #drop am feature
#     df = df.drop(['am'], axis=1)
    df=proc.get_past_n_steps(df, past_n_steps)
    #calculate forecast_time
    df['forecast_time'] = df['present_time']+timedelta(hours=pred_period)

    #join forecast according to forecast time 
    forecast = forecast.set_index('f_date') 
    forecast = forecast.add_suffix('_forecast')
    df = pd.merge(df, forecast, how = 'left', left_on = 'forecast_time', right_on ='f_date')
    #add cos day
    df = proc.smooth_day_hour(df)
    #fillna
    df.fillna(value=0, inplace=True)
    df = df.iloc[:-past_n_steps]
    #keep the latest information
    df_out = df.iloc[-1:]
    return df_out

## Check if any columns are missing:

In [16]:
# #to change: make an assert error
# df= prepare_x_test(measurement, forecast, 48, 6)
# #load model columns 
# model_cols = pickle.load(open('model_cols_past_48.pkl', 'rb'))
# print('columns missing:', model_cols.difference(df.columns))
# #it's okay if am feature is missing - it's not included in the model

In [17]:
df

speed_t-48  cos_wind_dir_t-48  sin_wind_dir_t-48  temp_t-48  \
280       5.545           0.596844          -0.766326  28.508333   

     radiation_t-48  precip_t-48  season_t-48  speed_t-47  cos_wind_dir_t-47  \
280      590.366667          0.0            3    4.923333           0.654128   

     sin_wind_dir_t-47  ...        present_time       forecast_time  \
280           -0.71467  ... 2020-07-14 16:00:00 2020-07-14 22:00:00   

     speed_forecast  temp_forecast  rad_forecast  precip_forecast  \
280             4.6           23.1           0.0              0.0   

     cos_wind_dir_forecast  sin_wind_dir_forecast  cos_day  cos_hour  
280                0.75471               0.656059   0.9711      -0.5  

[1 rows x 353 columns]

In [35]:
#predict next 48 hours
predict_list = np.arange(1,49,1)
prediction = pd.DataFrame(columns={})
for pred in predict_list:
    x_test = prepare_x_test(measurement, forecast, 48, int(pred))
    #drop timestamp columns and change to array format
    f_date = x_test['forecast_time'].iloc[0]
    p_date = x_test['present_time'].iloc[0]
    x_test = x_test.drop(['present_time','forecast_time'], axis=1)

    #check missing features 
    xgb1 = pickle.load(open('../trained_models/speed_t_'+str(pred), 'rb'))
    model_features = xgb1.get_booster().feature_names
    missing_features = set(model_features).difference(x_test.columns)
    if len(missing_features)>=1:
        print('model not run because some features are missing:', missing_features)
        break 
    else:
        #re-order features 
        x_test = x_test[model_features]
        
    #read 3 models for speed, cos_wind, sin_wind
    xgb1= pickle.load(open('../trained_models/speed_t_'+str(pred), 'rb'))
    xgb2 = pickle.load(open('../trained_models/cos_wind_dir_t_'+str(pred), 'rb'))
    xgb3 = pickle.load(open('../trained_models/sin_wind_dir_t_'+str(pred), 'rb'))

    #predict 
    speed = xgb1.predict(x_test)[-1]
    cos_wind = xgb2.predict(x_test)[-1]
    sin_wind = xgb3.predict(x_test)[-1]
    
    #calculate wind_dir
    wind_dir = utils.get_angle_in_degree(cos_wind, sin_wind)

    #record prediction
    prediction = prediction.append({'forecast_time':f_date,
                            'present_time':p_date,
                            'speed':speed,
                            'cos_wind_dir':cos_wind,
                           'sin_wind_dir':sin_wind,
                           'wind_dir':wind_dir}, ignore_index=True)


In [36]:
prediction

cos_wind_dir       forecast_time        present_time  sin_wind_dir  \
0       0.821470 2020-07-14 17:00:00 2020-07-14 16:00:00     -0.421536   
1       0.848212 2020-07-14 18:00:00 2020-07-14 16:00:00     -0.307416   
2       0.878282 2020-07-14 19:00:00 2020-07-14 16:00:00      0.014867   
3       0.832883 2020-07-14 20:00:00 2020-07-14 16:00:00      0.340638   
4       0.755188 2020-07-14 21:00:00 2020-07-14 16:00:00      0.619577   
5       0.726022 2020-07-14 22:00:00 2020-07-14 16:00:00      0.658296   
6       0.679650 2020-07-14 23:00:00 2020-07-14 16:00:00      0.739037   
7       0.491783 2020-07-15 00:00:00 2020-07-14 16:00:00      0.878513   
8       0.560999 2020-07-15 01:00:00 2020-07-14 16:00:00      0.815254   
9       0.657307 2020-07-15 02:00:00 2020-07-14 16:00:00      0.771324   
10      0.631006 2020-07-15 03:00:00 2020-07-14 16:00:00      0.793950   
11      0.548027 2020-07-15 04:00:00 2020-07-14 16:00:00      0.843435   
12      0.533386 2020-07-15 05:00:00 2020-07-14 16:00:00      0.813894   
13      0.527101 2020-07-15 06:00:00 2020-07-14 16:00:00      0.834918   
14      0.500898 2020-07-15 07:00:00 2020-07-14 16:00:00      0.838614   
15      0.508132 2020-07-15 08:00:00 2020-07-14 16:00:00      0.798602   
16      0.523736 2020-07-15 09:00:00 2020-07-14 16:00:00      0.730400   
17      0.619202 2020-07-15 10:00:00 2020-07-14 16:00:00      0.680011   
18      0.619618 2020-07-15 11:00:00 2020-07-14 16:00:00      0.466744   
19      0.747734 2020-07-15 12:00:00 2020-07-14 16:00:00      0.063178   
20      0.721158 2020-07-15 13:00:00 2020-07-14 16:00:00     -0.491220   
21      0.708315 2020-07-15 14:00:00 2020-07-14 16:00:00     -0.740020   
22      0.824275 2020-07-15 15:00:00 2020-07-14 16:00:00     -0.748488   
23      0.795512 2020-07-15 16:00:00 2020-07-14 16:00:00     -0.530847   
24      0.813738 2020-07-15 17:00:00 2020-07-14 16:00:00     -0.450908   
25      0.806122 2020-07-15 18:00:00 2020-07-14 16:00:00     -0.127342   
26      0.794170 2020-07-15 19:00:00 2020-07-14 16:00:00      0.203059   
27      0.611679 2020-07-15 20:00:00 2020-07-14 16:00:00      0.375191   
28      0.590954 2020-07-15 21:00:00 2020-07-14 16:00:00      0.607935   
29      0.559366 2020-07-15 22:00:00 2020-07-14 16:00:00      0.621280   
30      0.312438 2020-07-15 23:00:00 2020-07-14 16:00:00      0.685243   
31      0.506792 2020-07-16 00:00:00 2020-07-14 16:00:00      0.246439   
32      0.504761 2020-07-16 01:00:00 2020-07-14 16:00:00      0.469589   
33      0.481678 2020-07-16 02:00:00 2020-07-14 16:00:00      0.402745   
34      0.449124 2020-07-16 03:00:00 2020-07-14 16:00:00      0.497581   
35      0.481778 2020-07-16 04:00:00 2020-07-14 16:00:00      0.452956   
36      0.357115 2020-07-16 05:00:00 2020-07-14 16:00:00      0.304553   
37      0.469336 2020-07-16 06:00:00 2020-07-14 16:00:00      0.281570   
38      0.443908 2020-07-16 07:00:00 2020-07-14 16:00:00      0.417782   
39      0.465763 2020-07-16 08:00:00 2020-07-14 16:00:00      0.457129   
40      0.429586 2020-07-16 09:00:00 2020-07-14 16:00:00      0.386930   
41      0.459906 2020-07-16 10:00:00 2020-07-14 16:00:00      0.098219   
42      0.506963 2020-07-16 11:00:00 2020-07-14 16:00:00     -0.376823   
43      0.675919 2020-07-16 12:00:00 2020-07-14 16:00:00     -0.439671   
44      0.644510 2020-07-16 13:00:00 2020-07-14 16:00:00     -0.566147   
45      0.670147 2020-07-16 14:00:00 2020-07-14 16:00:00     -0.485653   
46      0.727966 2020-07-16 15:00:00 2020-07-14 16:00:00     -0.583983   
47      0.649497 2020-07-16 16:00:00 2020-07-14 16:00:00     -0.611627   

       speed    wind_dir  
0   4.762733  214.767743  
1   4.428719  211.982224  
2   4.418127   28.564237  
3   4.211710   33.603998  
4   4.114040   40.958196  
5   4.442183   43.446027  
6   4.086899   47.183732  
7   4.036884   60.542168  
8   3.343930   55.875114  
9   3.627290   48.905155  
10  3.392264   50.875634  
11  3.978087   56.768253  
12  3.536221   57.765451  

Hi Karim,

Please find my the updated notebook attached.\
I have made several changes to the code, and I will outline the changes made here: \
- crop out 7 hours of prediction data instead of 6
- when taking hourly average, now we are averaging to the 'right' instead of 'left' to reflect the hour reflect past hour data 
- xgb models are trained with column names. In addition, I've added several lines of code to compare model_features with testing_features, and the code will stop and raise an error flag of the missing features.

Models are all retrained to reflect the above amendments. I have pushed several directories to the github, and also sent the same files in a zip folder via Teams, including:
- trained_models: containing all the models in pickle files
- results: testing results (80% training / 20% testing data split) on the data set we have, to check performance after making the above 3 amendments. I have tested on max_depth=5, n_estimators = 100 and 200 respectively. Results are can be found in "xgboost_accuracy_in_48_depth_5_estim_100.csv" and "xgboost_accuracy_in_48_depth_5_estim_200.csv". I've compared this results with previously; now they are slightly worse (~3-4% decrease in accuracy), I think due to 1)croping out 1 more hour of forecast 2) shifting the hour to 1 hour later. Nevertheless, they are still good and consistent with accuracy at 83% for 3 hour prediction, 80% for 48 hour prediction.  

Please let me know if you have any problems with running the code, and good luck with the implementation process! 


Best,
Cynthia 